coding/python_style_guide.rst

```diff
@@ -12,648 +13,671 @@ To request changes to these standards, please file an :ref:`RFC <decision-making
+9. Public functions and classes
+10. Optional test suites
 
-   def proclist(alist=None):
-   if alist == None:
-   alist = []
+.. note:: Delete mention of test suites?
 
-   # if you can tolerate a tuple; tuples are immutable
-   def proclist(alist=()):
-       pass
+.. _style-guide-py-classes:
 
-Rather than the more obvious but dangerously wrong: 
+8. Classes
+==========
 
-.. code-block:: py
+Always decide whether a class's methods and instance variables (collectively: "attributes") should be public or non-public.
+If in doubt, choose non-public; it's easier to make it public later than to make a public attribute non-public.
 
-   def proclist(alist=[]):
-       pass
+Public attributes are those that you expect unrelated clients of your class to use, with your commitment to avoid backward incompatible changes.
+Non-public attributes are those that are not intended to be used by third parties; you make no guarantees that non-public attributes won't change or even be removed.
+
+We don't use the term "private" here, since no attribute is really private in Python (without a generally unnecessary amount of work).
+Another category of attributes are those that are part of the "subclass API" (often called "protected" in other languages).
+Some classes are designed to be inherited from, either to extend or modify aspects of the class's behavior.
+When designing such a class, take care to make explicit decisions about which attributes are public, which are part of the subclass API, and which are truly only to be used by your base class.
 
-.. _style-guide-py-7-3:
+For simple public data attributes, it is best to expose just the attribute name, without complicated accessor/mutator methods.
+Keep in mind that Python provides an easy path to future enhancement, should you find that a simple data attribute needs to grow functional behavior.
+In that case, use properties to hide functional implementation behind simple data attribute access syntax.
+
+- Note 1: Properties only work on new-style classes.
+
+- Note 2: Try to keep the functional behavior side-effect free, although side-effects such as caching are generally fine.
+
+- Note 3: Avoid using properties for computationally expensive operations; the attribute notation makes the caller believe that access is (relatively) cheap.
 
-Object type comparisons SHOULD always use ``isinstance()``
-----------------------------------------------------------
+.. _style-guide-py-super:
 
-Object type comparisons should always use :py:func:`isinstance()` instead of comparing types directly. 
+``super`` SHOULD NOT be used unless the author really understands the implications (e.g. in a well-understood multiple inheritance hierarchy).
+----------------------------------------------------------------------------------------------------------------------------------------------
+
+Python provides ``super`` so that each parent class' method is only called once (see https://www.python.org/download/releases/2.3/mro/).
```

Comment:
Surprisingly, not really. the 2.3 docs are the only ones with that page. https://docs.python.org/3/library/functions.html#super. But I suppose I could just link to Python 3's `super()` [docs](https://docs.python.org/3/library/functions.html#super) and Raymond Hettinger's [Super Considered Super!](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/) as a tonic to the blatant misinformation in https://fuhm.net/super-harmful/ that misled whoever wrote DM's original Python style guide.